frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+//! purchased by making a "valid call" dispatched through this pallet.
+//! By configuring certain calls to be valid for the lottery, you can encourage
+//! users to make those calls on your network. An example of how this could be
+//! used is to set validator nominations as a valid lottery call. If the lottery
+//! is set to repeat every month, then users would be encouraged to re-nominate
+//! validators every month.
+//!
+//! This pallet can be configured to use dynamically set calls or statically set
+//! calls. Call validation happens through the `ValidateCall` implementation.
+//! This pallet provides one implementation of this using the `CallIndices`
+//! storage item. You can also make your own implementation at the runtime level
+//! which can contain much more complex logic, such as validation of the
+//! parameters, which this pallet alone cannot do.
+//!
+//! This pallet uses the modulus operator to pick a random winner. It is known
+//! that this might introduce a bias if the random number chosen in a range that
+//! is not perfectly divisible by the total number of participants. The
+//! `MaxGenerateRandom` configuration can help mitigate this by generating new
+//! numbers until we hit the limit or we find a "fair" number. This is best
+//! effort only.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+mod benchmarking;
+pub mod weights;
+
+use sp_std::prelude::*;
+use sp_runtime::{DispatchError, ModuleId};
+use sp_runtime::traits::{AccountIdConversion, Saturating};
+use frame_support::{Parameter, decl_module, decl_error, decl_event, decl_storage, ensure, RuntimeDebug};
+use frame_support::dispatch::{Dispatchable, DispatchResult, GetDispatchInfo};
+use frame_support::traits::{
+	Currency, ReservableCurrency, Get, EnsureOrigin, ExistenceRequirement::KeepAlive, Randomness,
+};
+use frame_support::weights::Weight;
+use frame_system::ensure_signed;
+use codec::{Encode, Decode};
+pub use weights::WeightInfo;
+
+type BalanceOf<T> = <<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// The module's config trait.
+pub trait Config: frame_system::Config {
+	/// The Lottery's module id
+	type ModuleId: Get<ModuleId>;
+
+	/// A dispatchable call.
+	type Call: Parameter + Dispatchable<Origin=Self::Origin> + GetDispatchInfo + From<frame_system::Call<Self>>;
+
+	/// The currency trait.
+	type Currency: ReservableCurrency<Self::AccountId>;
+
+	/// Something that provides randomness in the runtime.
+	type Randomness: Randomness<Self::Hash>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+
+	/// The manager origin.
+	type ManagerOrigin: EnsureOrigin<Self::Origin>;
+
+	/// The max number of calls available in a single lottery.
+	type MaxCalls: Get<usize>;
+
+	/// Used to determine if a call would be valid for purchasing a ticket.
+	type ValidateCall: ValidateCall<Self>;
```

Comment:
the doc should say that it mustn't allow more than `MaxCalls` kind, where a kind of call is its first 2 bytes.
Because we use this assumption in benchmark